packages/core/src/new-api/internal/reconciliation/execution-state-resolver.ts

```diff
@@ -0,0 +1,161 @@
+import { IgnitionError } from "../../../errors";
+import { isRuntimeValue } from "../../type-guards";
+import {
+  DeploymentExecutionState,
+  ExecutionState,
+  ExecutionStateMap,
+  SendDataExecutionState,
+  StaticCallExecutionState,
+} from "../../types/execution-state";
+import {
+  AddressResolvableFuture,
+  ContractFuture,
+  Future,
+  ModuleParameterRuntimeValue,
+} from "../../types/module";
+import { isAddress } from "../utils";
+
+import { ReconciliationContext } from "./types";
+import { resolveModuleParameter, safeToString } from "./utils";
+
+// TODO: consider merging this into the execution state map
+export class ExecutionStateResolver {
+  // Library addresses are resolved from previous execution states
+  public static resolveLibrariesFromExecutionState(
+    libraries: Record<string, ContractFuture<string>>,
+    { executionStateMap }: ReconciliationContext
+  ): Record<string, string | undefined> {
+    return Object.fromEntries(
+      Object.entries(libraries).map(([key, libFuture]) => {
+        const executionStateEntry = executionStateMap[
+          libFuture.id
+        ] as DeploymentExecutionState;
+
+        return [key, executionStateEntry?.contractAddress];
+      })
+    );
+  }
+
+  public static resolveContractToAddress(
+    address: string | ContractFuture<string>,
+    { executionStateMap }: ReconciliationContext
+  ): string {
+    if (typeof address === "string") {
+      return address;
+    }
+
+    const contractAddress = this.resolveFromExecutionState(
+      address,
+      executionStateMap,
+      (exState: DeploymentExecutionState) => exState.contractAddress
+    );
+
+    if (contractAddress === undefined) {
+      throw new IgnitionError("Previous deployment without contractAddress");
```

Comment:
I think we should try to give more contextual information in these error messages. Is it added later? Otherwise this would be really hard to debug.

Reply:
I have changed a few of these to assertions (they should be picked up by validation or execution state shouldn't get in the state).

I have broadened out some of the errors with more info.

I also added a catch all for exceptions to `reconcileFutureSpecificReconciliations`. The only exceptions should be coming from the supporting reconciliation utility functions. The exception is now caught and returned as a reconciliation failure, but paired with the future's id for context.